dff/stats/savers/postgresql.py

```diff
@@ -0,0 +1,88 @@
+"""
+Postgresql Saver
+---------------------------
+Provides the Postgresql version of the :py:class:`~dff.stats.savers.saver.Saver`.
+The class should be constructed by calling the :py:func:`~dff.stats.savers.make_saver`
+factory with specific parameters.
+
+"""
+from typing import List
+from urllib import parse
+
+try:
+    from sqlalchemy.ext.asyncio import create_async_engine
+    from sqlalchemy import inspect, Table, MetaData, Column, String, Integer, JSON, DateTime, select, insert
+
+    IMPORT_ERROR_MESSAGE = None
+except ImportError as e:
+    IMPORT_ERROR_MESSAGE = e.msg
+
+from .saver import Saver
+from ..record import StatsRecord
+
+
+class PostgresSaver(Saver):
+    """
+    Saves the stats dataframe to - and reads from a Postgresql database.
+    The class should be constructed by calling the :py:func:`~dff.stats.savers.make_saver`
+    factory with specific parameters.
+
+    :param path: The construction path.
+        It should match the sqlalchemy :py:class:`~sqlalchemy.engine.Engine` initialization string.
+
+        .. code-block::
+
+            Saver("postgresql://user:password@localhost:5432/default")
+
+    :param table: Sets the name of the db table to use. Defaults to "dff_stats".
+    """
+
+    def __init__(self, path: str, table: str = "dff_stats") -> None:
+        if IMPORT_ERROR_MESSAGE is not None:
+            raise ImportError(IMPORT_ERROR_MESSAGE)
+        self.table = table
+        self.table_exists = False
+        parsed_path = parse.urlparse(path)
+        self.engine = create_async_engine(parse.urlunparse([(parsed_path.scheme + "+asyncpg"), *parsed_path[1:]]))
+        self.metadata = MetaData()
+        self.sql_table = Table(
+            self.table,
+            self.metadata,
+            Column("context_id", String),
+            Column("request_id", Integer),
+            Column("timestamp", DateTime),
+            Column("data_key", String),
+            Column("data", JSON),
+        )
+
+    async def save(self, data: List[StatsRecord]) -> None:
+        if not self.table_exists:
+            await self._create_table()
+            self.table_exists = True
+
+        async with self.engine.connect() as conn:
+            await conn.execute(insert(self.sql_table).values([item.dict() for item in data]))
+            await conn.commit()
+
+    async def load(self) -> List[StatsRecord]:
+        stats = []
+
+        async with self.engine.connect() as conn:
+            result = await conn.execute(select(self.sql_table))
```

Comment:
_create_table is asynchronous, so we cannot call it in __init__. As a fix, I moved the check to load as well. Since the flag is being set if the table exists, the database will only be queried once.

Reply:
Moved to the factory function